qiskit/qpy/binary_io/schedules.py

```diff
@@ -71,6 +76,29 @@ def _loads_symbolic_expr(expr_bytes):
     return expr
 
 
+def _format_legacy_qiskit_pulse_v5(pulse_type, envelope, parameters):
+    # In the transition to Qiskit Terra > 0.22, the representation of library pulses was changed from
+    # complex "amp" to float "amp" and "angle". To reflect this, QPY version was bumped to 6. The
+    # existing library pulses in QPY<=5 are handled here separately to conform with the new
+    # representation. To avoid role assumption for "amp" for custom pulses, only the library pulses
+    # are handled this way.
+
+    # List of pulses in the library in QPY version 5 and below:
+    v5_library_pulses = ["Gaussian", "GaussianSquare", "Drag", "Constant"]
+
+    if pulse_type in v5_library_pulses:
+        if isinstance(
+            parameters["amp"], complex
```

Comment:
With terra <0.22 amp is "always" complex because of [this line](https://github.com/Qiskit/qiskit-terra/blob/b82a89d920851292152601b5c14367744c2244bd/qiskit/pulse/library/symbolic_pulses.py#L436-L437). This implies the formatted angle parameter becomes always 0.0 (float) due to the return type of `np.angle`, which causes the dict inequality since default angle parameter is 0 (integer). 

You need to use `if parameters["amp"].imag == 0:` or remove [this branching](https://github.com/Qiskit/qiskit-terra/blob/54bcdf28883689047599581be340fcdc6f622faf/qiskit/pulse/library/symbolic_pulses.py#L925-L926). I would do the latter one, because having complex amplitude with `angle=0.0` doesn't break backward compatibility (it still returns complex amplitude -- let me know if I am missing something). In the current logic, terra 0.23 will have two envelopes `_amp` and `_amp * (sym.I * _angle)`, and we will need to write another formatter for this particular version (because terra 0.25 we may drop complex amplitude and always use `_amp * (sym.I * _angle)`).